torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1976,6 +1976,119 @@ def print_results(grouped, indent, result_fn):
 
 TLS = threading.local()
 
+# Event class for reconstructing ShapeEnv at arbitrary time.
+#
+# Represents a method call that mutates ShapeEnv in a way that affects the
+# issued guards, when ShapeEnv.produce_guards is called.
+@dataclass
+class ShapeEnvEvent:
+    # ShapeEnv method.
+    f: Callable
+
+    # Arguments and keyword arguments called with.
+    args: Optional[List[Any]] = None
+    kwargs: Optional[Dict[str, Any]] = None
+
+    # Length of tracked_fakes at the time the method was called.
+    tracked_fakes_length: Optional[int] = None
+
+    # Name of the captured event.
+    # Used for special handling of particular methods.
+    name: Optional[str] = None
+
+    # Replay itself, but using shape_env as self.
+    #
+    # mapping: an optional FX node converter. This is used for converting
+    # the captured FX nodes from the old ShapeEnv by corresponding ones from
+    # shape_env.
+    def __call__(
+        self,
+        shape_env: Optional["ShapeEnv"] = None,
+        mapping: Optional[Callable[[torch.fx.Node], torch.fx.Node]] = None
+    ) -> Any:
+        # Special handling for the constructor event.
+        if self.f == ShapeEnv:
+            assert shape_env is None and self.args is None and self.kwargs is not None
+            return ShapeEnv(**self.kwargs)
+
+        assert shape_env is not None
+
+        args = list(self.args or list())
+        kwargs = dict(self.kwargs or dict())
+
+        # Converts FX nodes using the mapping argument.
+        def convert_node(x: Any) -> Any:
+            return x if mapping is None or not isinstance(x, torch.fx.Node) else mapping(x)
+
+        # Replaces the value of an specific argument by the result of fn.
+        def replacearg(index: int, key: str, fn: Callable):
+            if index < len(args):
+                args[index] = fn(args[index])
+            if key in kwargs:
+                kwargs[key] = fn(kwargs[key])
+
+        if self.is_create_fx_call_function():
+            # ShapeEnv.create_fx_call_function:
+            # "args" parameter is a tuple of FX nodes from the FX graph of the old ShapeEnv.
+            # They must be replaced, since a "call_function" FX node with this tuple as argument
+            # will be added to the FX graph of the new shape_env.
+            replacearg(index=1, key="args", fn=lambda args: tuple(convert_node(a) for a in args))
+        if self.is_evaluate_expr():
+            # ShapeEnv.evaluate_expr:
+            # "fx_node" parameter is an (optional) FX node that represents the evaluate expression.
+            # They must be replaced, since it will be part of a "call_function" FX node for
+            # torch._assert, which will be added to the FX graph of the new shape_env.
+            replacearg(index=3, key="fx_node", fn=convert_node)
+
+        # Actually call the method with the converted arguments.
+        return self.f(shape_env, *args, **kwargs)
```

Comment:
This seems pretty delicate; in particular, it seems difficult to guarantee that the replay mechanism is going to keep working as people keep monkeying around with methods on ShapeEnv. First off, is this well tested? Second, one way to make it structurally harder to "mess it up" is to force all substantive method calls through this API, so that even regular, non bisecting ShapeEnv calls go through the logic here.